src/sources/http_scrape/scrape.rs

```diff
@@ -0,0 +1,252 @@
+#[async_trait::async_trait]
+#[typetag::serde(name = "http_scrape")]
+impl SourceConfig for HttpScrapeConfig {
+    async fn build(&self, cx: SourceContext) -> Result<sources::Source> {
+        // build the url
+        let endpoints = vec![self.endpoint.clone()];
+        let urls = endpoints
+            .iter()
+            .map(|s| s.parse::<Uri>().context(sources::UriParseSnafu))
+            .map(|r| r.map(|uri| build_url(&uri, &self.query)))
+            .collect::<std::result::Result<Vec<Uri>, sources::BuildError>>()?;
+
+        let tls = TlsSettings::from_options(&self.tls)?;
+
+        let log_namespace = cx.log_namespace(self.log_namespace);
+
+        // build the decoder
+        let decoder =
+            DecodingConfig::new(self.framing.clone(), self.decoding.clone(), log_namespace).build();
+
+        let content_type = self.decoding.content_type(&self.framing).to_string();
+
+        // the only specific context needed is the codec decoding
+        let context = HttpScrapeContext {
+            decoder,
+            log_namespace,
+        };
+
+        let inputs = GenericHttpScrapeInputs {
+            urls,
+            interval_secs: self.scrape_interval_secs,
+            headers: self.headers.clone(),
+            content_type,
+            auth: self.auth.clone(),
+            tls,
+            proxy: cx.proxy.clone(),
+            shutdown: cx.shutdown,
+        };
+
+        Ok(http_scrape(inputs, context, cx.out).boxed())
+    }
+
+    fn outputs(&self, global_log_namespace: LogNamespace) -> Vec<Output> {
+        // There is a global and per-source `log_namespace` config. The source config overrides the global setting,
+        // and is merged here.
+        let log_namespace = global_log_namespace.merge(self.log_namespace);
+
+        let schema_definition = self
+            .decoding
+            .schema_definition(log_namespace)
+            .with_standard_vector_source_metadata();
+
+        vec![Output::default(self.decoding.output_type()).with_schema_definition(schema_definition)]
+    }
+
+    fn source_type(&self) -> &'static str {
+        NAME
+    }
+
+    fn can_acknowledge(&self) -> bool {
+        false
+    }
+}
+
+#[derive(Clone)]
+struct HttpScrapeContext {
+    decoder: Decoder,
+    log_namespace: LogNamespace,
+}
+
+impl HttpScrapeContext {
+    /// Decode the events from the byte buffer
+    fn decode_events(&mut self, buf: &mut BytesMut) -> Vec<Event> {
+        let mut events = Vec::new();
+        loop {
+            match self.decoder.decode_eof(buf) {
+                Ok(Some((next, _))) => {
+                    events.extend(next.into_iter());
+                }
+                Ok(None) => break,
+                Err(error) => {
+                    // Error is logged by `crate::codecs::Decoder`, no further
+                    // handling is needed here.
+                    if !error.can_continue() {
+                        break;
+                    }
+                    break;
```

Comment:
Is there a path here that _doesn't_ `break`? This looks redundant.

Reply:
Good catch. Yes I'll remove that.